File: Test Case Generation/gemini/cpp/test_task_5.cpp
#include <gtest/gtest.h>
#include <vector>

using namespace std;

// Função original integrada para garantir a compilação nos testes
vector<int> intersperse(vector<int> numbers, int delimeter) { 
    vector<int> out = {};
    if (numbers.size() > 0) out.push_back(numbers[0]);
    for (int i = 1; i < (int)numbers.size(); i++) {
        out.push_back(delimeter);
        out.push_back(numbers[i]);
    }
    return out;
}

// Teste para vetor vazio (Edge Case: size == 0)
TEST(IntersperseTest, EmptyVector) {
    vector<int> input = {};
    vector<int> expected = {};
    EXPECT_EQ(intersperse(input, 4), expected);
}

// Teste para vetor com um único elemento (Edge Case: loop não deve executar)
TEST(IntersperseTest, SingleElement) {
    vector<int> input = {1};
    vector<int> expected = {1};
    EXPECT_EQ(intersperse(input, 4), expected);
}

// Teste para vetor com múltiplos elementos (Caminho feliz / Line coverage do loop)
TEST(IntersperseTest, MultipleElements) {
    vector<int> input = {1, 2, 3};
    vector<int> expected = {1, 4, 2, 4, 3};
    EXPECT_EQ(intersperse(input, 4), expected);
}

// Teste com o delimitador sendo zero ou negativo
TEST(IntersperseTest, ZeroAndNegativeDelimeter) {
    vector<int> input = {5, 5};
    vector<int> expected_zero = {5, 0, 5};
    vector<int> expected_neg = {5, -1, 5};
    
    EXPECT_EQ(intersperse(input, 0), expected_zero);
    EXPECT_EQ(intersperse(input, -1), expected_neg);
}

// Teste com elementos repetidos no input
TEST(IntersperseTest, RepeatedElements) {
    vector<int> input = {1, 1, 1};
    vector<int> expected = {1, 2, 1, 2, 1};
    EXPECT_EQ(intersperse(input, 2), expected);
}

// Teste de integridade de grandes valores
TEST(IntersperseTest, LargeValues) {
    vector<int> input = {2147483647, 2147483647};
    vector<int> expected = {2147483647, 100, 2147483647};
    EXPECT_EQ(intersperse(input, 100), expected);
}

File: Test Case Generation/gemini/python/test_task_5.py
import pytest
from typing import List

# Assumindo que a função está no mesmo ficheiro ou importada corretamente
# from modulo_original import intersperse

def intersperse(numbers: List[int], delimeter: int) -> List[int]:
    """ Insert a number 'delimeter' between every two consecutive elements of input list `numbers'
    >>> intersperse([], 4)
    []
    >>> intersperse([1, 2, 3], 4)
    [1, 4, 2, 4, 3]
    """
    if not numbers:
        return []

    result = []

    for n in numbers[:-1]:
        result.append(n)
        result.append(delimeter)

    result.append(numbers[-1])

    return result

@pytest.mark.parametrize("numbers, delimeter, expected", [
    # Teste de lista vazia (Edge Case: Input vazio)
    ([], 4, []),
    # Teste com apenas um elemento (Edge Case: Sem pares para intercalar)
    ([1], 4, [1]),
    # Teste com dois elementos
    ([1, 2], 4, [1, 4, 2]),
    # Teste padrão do docstring
    ([1, 2, 3], 4, [1, 4, 2, 4, 3]),
    # Teste com números negativos
    ([-1, -2], -3, [-1, -3, -2]),
    # Teste com zeros
    ([0, 0, 0], 0, [0, 0, 0, 0, 0]),
    # Teste com lista grande
    (list(range(5)), 99, [0, 99, 1, 99, 2, 99, 3, 99, 4]),
    # Teste com delimitador sendo o mesmo que os números
    ([1, 1, 1], 1, [1, 1, 1, 1, 1]),
])
def test_intersperse_scenarios(numbers: List[int], delimeter: int, expected: List[int]):
    """Valida múltiplos cenários para garantir 100% de cobertura de linha e lógica."""
    assert intersperse(numbers, delimeter) == expected

def test_intersperse_immutability():
    """Garante que a lista original não é modificada (Side effect check)."""
    original = [1, 2, 3]
    copy_original = original.copy()
    intersperse(original, 4)
    assert original == copy_original

def test_intersperse_with_large_integers():
    """Teste com inteiros de grande magnitude."""
    assert intersperse([10**18, 10**18], 1) == [10**18, 1, 10**18]